linera-chain/src/data_types.rs

```diff
@@ -692,13 +689,19 @@ impl ExecutedBlock {
     ) -> Option<MessageId> {
         let block = &self.block;
         let transaction_index = block.incoming_messages.len().checked_add(operation_index)?;
-        let first_message_index = match transaction_index.checked_sub(1) {
-            None => 0,
-            Some(index) => *self.outcome.message_counts.get(index)?,
-        };
+        let first_message_index = u32::try_from(
+            self.outcome
+                .messages
+                .iter()
+                .take(transaction_index)
+                .map(Vec::len)
+                .sum::<usize>(),
+        )
+        .ok()?;
         let index = first_message_index.checked_add(message_index)?;
```

Comment:
nit: move this line to 704

Reply:
(I moved the message index check up instead.)